PSReadLine/WindowsKeyboardLayoutUtil.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.PowerShell
+{
+    internal static class WindowsKeyboardLayoutUtil
+    {
+        /// <remarks>
+        /// <para>
+        ///     This method helps to find the active keyboard layout in a terminal process that controls the current
+        ///     console application. The terminal process is not always the direct parent of the current process, but
+        ///     may be higher in the process tree in case PowerShell is a child of some other console process.
+        /// </para>
+        /// <para>
+        ///     Currently, we check up to 20 parent processes to see if their main window (as determined by the
+        ///     <see cref="Process.MainWindowHandle"/>) is visible.
+        /// </para>
+        /// <para>
+        ///     If this method returns <c>null</c>, it means it was unable to find the parent terminal process, and so
+        ///     you have to call the <see cref="GetConsoleKeyboardLayoutFallback"/>, which is known to not work properly
+        ///     in certain cases, as documented by https://github.com/PowerShell/PSReadLine/issues/1393
+        /// </para>
+        /// </remarks>
+        public static IntPtr? GetConsoleKeyboardLayout()
+        {
+            // Define a limit not get stuck in case processed form a loop (possible in case pid reuse).
+            const int iterationLimit = 20;
+
+            var pbi = new PROCESS_BASIC_INFORMATION();
+            var process = Process.GetCurrentProcess();
+            for (var i = 0; i < iterationLimit; ++i)
+            {
+                var isVisible = IsWindowVisible(process.MainWindowHandle);
+                if (!isVisible)
+                {
+                    // Main process window is invisible. This is not (likely) a terminal process.
+                    var status = NtQueryInformationProcess(process.Handle, 0, ref pbi, Marshal.SizeOf(pbi), out var _);
+                    if (status != 0 || pbi.InheritedFromUniqueProcessId == IntPtr.Zero)
+                        break;
+
+                    try
+                    {
+                        process = Process.GetProcessById(pbi.InheritedFromUniqueProcessId.ToInt32());
+                    }
+                    catch (Exception)
+                    {
+                        // No access to the process, or the process is already dead. Either way, we cannot determine its
+                        // keyboard layout.
+                        return null;
+                    }
+
+                    continue;
+                }
+
+                var tid = GetWindowThreadProcessId(process.MainWindowHandle, out _);
+                if (tid == 0) return null;
+                return GetKeyboardLayout(tid);
+            }
+
+            return null;
```

Comment:
Instead of this, how about just return `GetKeyboardLayout(0)` here?
```suggestion
            // comment like "fall back to the default keyboard when we failed to find the parent terminal process"
            return GetKeyboardLayout(0);
```